samples/aspnetcore/blazor/Validation/Pages/Index.razor

```diff
@@ -0,0 +1,49 @@
+﻿@page "/"
+@using Validation.Data
+
+<h1>Validation of nested objects</h1>
+
+<EditForm Model="@person" OnValidSubmit="@OnValidSubmit" OnInvalidSubmit="@OnInvalidSubmit">
+    <RecursiveDataAnnotationsValidator />
+    <ValidationSummary />
+
+    <p>
+        Name: <InputText @bind-Value="person.Name" />
+    </p>
+    <p>
+        Street: <InputText @bind-Value="person.Address.Street" />
+    </p>
+    <p>
+        City: <InputText @bind-Value="person.Address.City" />
+    </p>
+
+    @foreach (var number in person.PhoneNumbers)
+    {
+        <p>
+            Phone: <InputText @bind-Value="number.Number" />
```

Comment:
Maybe consider replacing the `<ValidationSummary />` with a set of `<ValidationMessage />` components (one next to each input). Otherwise if you get an invalid phone number, for example, the message won't make clear *which* of the phone numbers is invalid.